frame/dex/src/lib.rs

```diff
@@ -524,14 +507,59 @@ pub mod pallet {
 			to: &T::AccountId,
 			amount: AssetBalanceOf<T>,
 			keep_alive: bool,
-		) -> Result<<T as pallet::Config>::AssetBalance, DispatchError> {
+		) -> Result<T::AssetBalance, DispatchError> {
+			Self::deposit_event(Event::Transfer {
+				from: from.clone(),
+				to: to.clone(),
+				asset: asset_id,
+				amount,
+			});
 			match asset_id {
 				MultiAssetId::Native => T::Currency::transfer(from, to, amount, keep_alive),
 				MultiAssetId::Asset(asset_id) =>
 					T::Assets::transfer(asset_id, from, to, amount, keep_alive),
 			}
 		}
 
+		fn do_swap(
+			sender: &T::AccountId,
+			amounts: &Vec<AssetBalanceOf<T>>,
+			path: &BoundedVec<MultiAssetId<T::AssetId>, T::MaxSwapPathLength>,
+			send_to: &T::AccountId,
+			keep_alive: bool,
+		) -> Result<(), DispatchError> {
+			if let Some(&[asset1, asset2]) = path.get(0..2) {
+				let pool_id = Self::get_pool_id(asset1, asset2);
+				let pool_account = Self::get_pool_account(pool_id);
+				let first_amount = amounts.first().expect("Always has more than one element");
+
+				Self::transfer(asset1, sender, &pool_account, *first_amount, keep_alive)?;
+
+				let path_len = path.len() as u32;
+				for i in 0..(path_len - 1) {
+					if let Some(&[asset1, asset2]) = path.get(i as usize..(i + 2) as usize) {
+						let pool_id = Self::get_pool_id(asset1, asset2);
+						let (sorted_asset1, sorted_asset2) = pool_id;
+						let pool_account = Self::get_pool_account(pool_id);
+
+						let amount_out =
+							amounts.get((i + 1) as usize).ok_or(Error::<T>::PathError)?;
+						let to = if i < path_len - 2 {
+							let asset3 = path.get((i + 2) as usize).ok_or(Error::<T>::PathError)?;
+							Self::get_pool_account(Self::get_pool_id(asset2, *asset3))
+						} else {
+							send_to.clone()
+						};
+
+						let send_asset =
+							if asset1 == sorted_asset1 { sorted_asset2 } else { sorted_asset1 };
```

Comment:
I assert send_asset == asset2 is always true here.

Reply:
nope, you can provide the path as `[asset1, asset2]` or as `[asset2, asset1]`

Reply:
but when get reserves gets called whatever it will return the reserves the right way around. The only thing that is possible to not be the right way around is the `(sorted_asset1, sorted_asset2)`.

Reply:
you're right, fixed